lib/puppet/functions/vault_lookup/lookup.rb

```diff
@@ -2,48 +2,77 @@
     end
 
-    uri = URI(vault_url)
-    # URI is used here to just parse the vault_url into a host string
+    vault_base_uri = URI(vault_url)
+    # URI is used here to parse the vault_url into a host string
     # and port; it's possible to generate a URI::Generic when a scheme
     # is not defined, so double check here to make sure at least
     # host is defined.
-    raise Puppet::Error, "Unable to parse a hostname from #{vault_url}" unless uri.hostname
+    raise Puppet::Error, "Unable to parse a hostname from #{vault_url}" unless vault_base_uri.hostname
 
-    use_ssl = uri.scheme == 'https'
-    connection = Puppet::Network::HttpPool.http_instance(uri.host, uri.port, use_ssl)
+    client = Puppet.runtime[:http]
+    token = get_cert_auth_token(client,
+                                vault_base_uri,
+                                vault_namespace,
+                                vault_cert_path_segment,
+                                vault_cert_role)
 
-    token = get_auth_token(connection)
+    secret_uri = vault_base_uri + "/v1/#{path.delete_prefix('/')}"
+    data = get_secret(client, secret_uri, token)
+    Puppet::Pops::Types::PSensitiveType::Sensitive.new(data)
+  end
 
-    secret_response = connection.get("/v1/#{path}", 'X-Vault-Token' => token)
-    unless secret_response.is_a?(Net::HTTPOK)
-      message = "Received #{secret_response.code} response code from vault at #{uri.host} for secret lookup"
-      raise Puppet::Error, append_api_errors(message, secret_response)
+  private
+
+  def auth_login_body(vault_cert_role)
+    if vault_cert_role
+      { name: vault_cert_role }.to_json
+    else
+      ''
     end
+  end
 
+  def get_secret(client, uri, token)
+    secret_response = client.get(uri, headers: { 'X-Vault-Token' => token })
+    unless secret_response.success?
+      message = "Received #{secret_response.code} response code from vault at #{uri} for secret lookup"
+      raise Puppet::Error, append_api_errors(message, secret_response)
+    end
     begin
-      data = JSON.parse(secret_response.body)['data']
+      JSON.parse(secret_response.body)['data']
     rescue StandardError
       raise Puppet::Error, 'Error parsing json secret data from vault response'
     end
-
-    Puppet::Pops::Types::PSensitiveType::Sensitive.new(data)
   end
 
-  private
-
-  def get_auth_token(connection)
-    response = connection.post('/v1/auth/cert/login', '')
-    unless response.is_a?(Net::HTTPOK)
-      message = "Received #{response.code} response code from vault at #{connection.address} for authentication"
+  def get_cert_auth_token(client, vault_url, vault_namespace, vault_cert_path_segment, vault_cert_role)
+    role_data = auth_login_body(vault_cert_role)
+    headers = { 'Content-Type' => 'application/json' }
+    if vault_namespace
+      headers['X-Vault-Namespace'] = vault_namespace
+    end
+    login_url = vault_url + vault_cert_path_segment + 'login'
```

Comment:
By default `vault_cert_path_segment` ends with '/', so the above concatenation works as expected:

```ruby
irb(main):016:0> vault_url + 'v1/auth/cert/' + 'login'
=> #<URI::HTTPS https://vault.example.com/v1/auth/cert/login>
```

But if the trailing slash is omitted, the results are surprising:

```
irb(main):017:0> vault_url + 'v1/auth/cert' + 'login'
=> #<URI::HTTPS https://vault.example.com/v1/auth/login>
```

Maybe append the trailing slash if it's missing?